File: fathersDay/Dsa/Day7 (121. Best Time to Buy and Sell Stock)/Solution.cpp
#include <vector>
#include <algorithm> // For std::min and std::max

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }
        
        int min_price = prices[0]; // Initialize with the first price
        int max_profit = 0;
        
        // Iterate from the second element
        for (int i = 1; i < prices.size(); ++i) {
            // Update min_price if a lower price is found
            min_price = std::min(min_price, prices[i]);
            
            // Calculate current profit and update max_profit if it's higher
            max_profit = std::max(max_profit, prices[i] - min_price);
        }
        
        return max_profit;
    }
};

File: fathersDay/Dsa/Day7 (121. Best Time to Buy and Sell Stock)/Solution.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        
        int minPrice = prices[0]; // Initialize with the first price
        int maxProfit = 0;
        
        // Iterate from the second element
        for (int i = 1; i < prices.length; i++) {
            // Update minPrice if a lower price is found
            minPrice = Math.min(minPrice, prices[i]);
            
            // Calculate current profit and update maxProfit if it's higher
            maxProfit = Math.max(maxProfit, prices[i] - minPrice);
        }
        
        return maxProfit;
    }
}